process/block/metablockProposal.go

```diff
@@ -402,3 +499,209 @@ func (mp *metaProcessor) selectIncomingMiniBlocks(
+	if err != nil {
+		return fmt.Errorf("%w : checkShardHeadersValidityAndFinalityProposal -> verifyUsedShardHeadersValidity", err)
+	}
+
+	return mp.checkShardInfoValidity(metaHeaderHandler, usedShardHeadersInfo)
+}
+
+func (mp *metaProcessor) checkShardInfoValidity(metaHeaderHandler data.MetaHeaderHandler, usedShardHeadersInfo *UsedShardHeadersInfo) error {
+	createdShardInfoProposal, createdShardInfo, err := mp.shardInfoCreateData.CreateShardInfoV3(metaHeaderHandler, usedShardHeadersInfo.OrderedShardHeaders, usedShardHeadersInfo.OrderedShardHeaderHashes)
+	if err != nil {
+		return fmt.Errorf("%w : checkShardInfoValidity -> CreateShardInfoV3", err)
+	}
+
+	headerShardInfo := metaHeaderHandler.GetShardInfoHandlers()
+	headerShardInfoProposal := metaHeaderHandler.GetShardInfoProposalHandlers()
+	if len(createdShardInfo) != len(headerShardInfo) || len(createdShardInfoProposal) != len(headerShardInfoProposal) {
+		return process.ErrHeaderShardDataMismatch
+	}
+
+	for i := 0; i < len(headerShardInfo); i++ {
+		if !headerShardInfo[i].Equal(createdShardInfo[i]) {
+			return fmt.Errorf("%w for shardInfo item %d", process.ErrHeaderShardDataMismatch, i)
+		}
+	}
+	for i := 0; i < len(headerShardInfoProposal); i++ {
+		if !headerShardInfoProposal[i].Equal(createdShardInfoProposal[i]) {
+			return fmt.Errorf("%w for shardInfoProposal item %d", process.ErrHeaderShardDataMismatch, i)
+		}
+	}
+
+	return nil
+}
+
+func (mp *metaProcessor) verifyUsedShardHeadersValidity(
+	usedShardHeaders map[uint32][]ShardHeaderInfo,
+	lastCrossNotarizedHeader map[uint32]ShardHeaderInfo,
+) error {
+	var err error
+	for shardID, hdrsForShard := range usedShardHeaders {
+		lastNotarizedHeaderInfoForShard := lastCrossNotarizedHeader[shardID]
+		for _, shardHdrInfo := range hdrsForShard {
+			if !mp.isGenesisShardBlockAndFirstMeta(shardHdrInfo.Header.GetNonce()) {
+				err = mp.headerValidator.IsHeaderConstructionValid(shardHdrInfo.Header, lastNotarizedHeaderInfoForShard.Header)
+				if err != nil {
+					return err
+				}
+			}
+
+			lastNotarizedHeaderInfoForShard = shardHdrInfo
+		}
+	}
+	return nil
+}
+
+func (mp *metaProcessor) HasProofsForHeaders(headersPerShard map[uint32][]ShardHeaderInfo) bool {
+	for _, headersForShard := range headersPerShard {
+		for _, headerInfo := range headersForShard {
+			if !mp.proofsPool.HasProof(headerInfo.Header.GetShardID(), headerInfo.Hash) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (mp *metaProcessor) getShardHeadersFromMetaHeader(
+	metaHeaderHandler data.MetaHeaderHandler,
+) (*UsedShardHeadersInfo, error) {
+	shardInfoProposalHandlers := metaHeaderHandler.GetShardInfoProposalHandlers()
+	usedShardHeaders := make(map[uint32][]ShardHeaderInfo)
+	var err error
+	var header data.HeaderHandler
+	orderedShardHeaders := make([]data.HeaderHandler, 0, len(shardInfoProposalHandlers))
+	orderedShardHeaderHashes := make([][]byte, 0, len(shardInfoProposalHandlers))
```

Comment:
These are not actually ordered, since after calling this func, you do this:
```
for _, headers := range usedShardHeadersInfo.HeadersPerShard {
		sort.Slice(headers, func(i, j int) bool {
			return headers[i].Header.GetNonce() < headers[j].Header.GetNonce()
		})
	}
```
1. Perhaps we could move the sorting here?
2.  Shouldn't we also sort the hashes?